(#1694605) Backport/fixes for issues found by LGTM on RHEL 7

mrc0mmand opened the pull request:
This PR attempts to fix errors & warnings found by LGTM. Recommendations are currently ignored, as it would make this PR unnecessarily big.

**Note:** some of these patches are not even in upstream (mostly because the particular code was completely rewritten in upstream).

Unresolved errors/warnings:
 - **[error]** `src/shared/pager.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/fe7b6b6c19424e12d6fb721e94422a5ff28b7125/files/src/shared/pager.c?sort=name&dir=ASC&mode=heatmap#x437173520f29364c:1) - pager-related stuff has been significantly refactored in upstream, needs further investigation
 - ~~*[ignored]* `src/libsystemd/sd-bus/bus-bloom.h` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/libsystemd/sd-bus/bus-bloom.c?sort=name&dir=ASC&mode=heatmap) - dropped from upstream along with kdbus~~
 - `src/shared/strv.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/shared/strv.c?sort=name&dir=ASC&mode=heatmap#x90c37d31240b843a:1) - `strv_join_quoted` was dropped from upstream (replaced by a different approach) without resolving this issue; would require some, possibly substantial, refactoring
 - ~~*[ignored]* `src/bootchart/*` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/bootchart/bootchart.c#x9c7e30d234d48467:1) - dropped from upstream (systemd/systemd@232c84b2d22f2d96982b3c0390d29498bb430835)~~
 - `src/core/execute.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/core/execute.c?sort=name&dir=ASC&mode=heatmap#xbfa2ae16fedaa55f:1) - `process_execute` has been refactored to support longer process names (systemd/systemd@9bfaffd5a9c138b0bbdfb452b3f4435746f2d18b), not sure if it's worth backporting
 - ~~*[ignored]* `src/gudev/*` - dropped from upstream (systemd/systemd@2375607039517c88df51ef16ddbb624ec1c10654)~~
 - `src/shared/fw-util.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/shared/fw-util.c#x62def701dfc13e50:1) - renamed to `src/shared/firewall-util.c`, however, not yet fixed in upstream

The *[ignored]* rules should *hopefully* take effect once this PR is merged.

keszybz committed:
systemctl: remove dead check

r could never be less than zero.

CID #1271350.

(cherry picked from commit 2558691285b2f7b3078b43b26736fcadf3c6d194)

Resolves: #1694605

ssahani committed:
lldp: set correct state for processing

ldp_receive_frame after correct processing of the packet the state
should be LLDP_AGENT_RX_WAIT_FOR_FRAME not LLDP_AGENT_RX_UPDATE_INFO.

(cherry picked from commit 9bb1bffe3018aef5e70c411f68863eafdc92fddf)

Resolves: #1694605

poettering committed:
timedate: use gmtime_r() and localtime_r()

gmtime() and localtime() operate on a static buffer. let's avoid this,
as we never know whether some library might use these calls in some
backrgound thread.

Discovered by lgtm:

https://lgtm.com/projects/g/systemd/systemd/
(cherry picked from commit e46acb7950a9f07ac60d772309de842c444ad2bd)

Resolves: #1694605

evverx committed:
sd-bus: stop using the result of an assignment as an operand of &&

This makes OBJECT_PATH_FOREACH_PREFIX consistent with PATH_FOREACH_PREFIX
and also fixes 7 alerts reported by LGTM at
https://lgtm.com/projects/g/systemd/systemd/snapshot/ac0a08700344a5690803df8dd80e8bb5013184a5/files/src/libsystemd/sd-bus/bus-objects.c?sort=name&dir=ASC&mode=heatmap&showExcluded=true#V1383

(cherry picked from commit 283c2653877f2b958045ad99493869fe1a88be95)

Resolves: #1694605

yuwata committed:
sd-bus: fix implicit downcast of bitfield reported by LGTM

(cherry picked from commit affaed1e55685e8e248db1afd50b3e0f466ce5cc)

Resolves: #1694605

yuwata committed:
analyze: actually select longest activated-time of services

Discovered by LGTM.

(cherry picked from commit e68131526b7a2cfdb8e59bd2ed7a284dba095ac5)

Resolves: #1694605

yuwata committed:
journal-gateway: explicitly declare local variables

Suggested by LGTM.

(cherry picked from commit c497e449f41774a36e01ae2cc2abade6133dffe1)

Resolves: #1694605

yuwata committed:
journal-gateway: use localStorage["cursor"] only when it has valid value

Discovered by LGTM.

(cherry picked from commit 944072feddb73333023d0a98bf87fd2a17f894d3)

Resolves: #1694605

keszybz committed:
journal-verify: add comment and silence LGTM warning

(cherry picked from commit 0a587335d2a16a48479b10969491a7353d78c25b)

Resolves: #1694605

keszybz committed:
journal-upload: add asserts that snprintf does not return an error

LGMT complains:
> The size argument of this snprintf call is derived from its return value,
> which may exceed the size of the buffer and overflow.

Let's make sure that r is non-negative. (This shouldn't occur unless the format
string is borked, so let's just add an assert.)
Then, let's reorder the comparison to avoid the potential overflow.

(cherry picked from commit 91db8ed5b2e67abf738381a6ed6a05a8271498cd)

Resolves: #1694605

lnykryn commented:
This pull request **fixes 13 alerts** when merging 4df22c749bd91260f314bb94ad38ad3e735d368f into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-627246f7bbd1809db84daf05eea60f337f32a9e5)

**fixed alerts:**

* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand committed:
pid1: remove unnecessary error reassignment

r is always >= 0 so let's drop the unnecessary condition
to make LGTM happy

Based on commit 7c3733d5defb5428ddfeefae6d80fcbcc253e7ef

Resolves: #1694605

mrc0mmand committed:
core: drop unnecessary condition

n is always > 0 as all goto statements for this section are already
guarded by a (n > 0) condition. Let's drop it from the fail section
to make LGTM happy.

Relevant commits:
11a30cec2a9b6168b024c06720ad238dd1390794
59eeb84ba65483c5543d1bc840c2ac75642ef638

Resolves: #1694605

mrc0mmand committed:
core: drop unnecessary condition

In this case ret is always >= 0 as it still contains the initialization
value (i.e. 0). Let's drop the unnecessary condition to make LGTM happy.

Relevant commit: efdb02375beb0a940c3320865572913780b4d7de

Resolves: #1694605

mrc0mmand committed:
import: drop unnecessary condition

Relevant commit: d61b34f1b22d1d7a641a1652ff23113d3e87ace7

Resolves: #1694605

lnykryn commented:
This pull request **fixes 15 alerts** when merging 73386a79dec0cfd5a9341f7b4c95e82fb98da1a0 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-7fd5027e09ab908560398da1bb228b78e183331e)

**fixed alerts:**

* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 2 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand committed:
journal: drop redundant condition

Relevant commit: 29bfb683a8c9d03f10db58b4e61351fd9048f270

Resolves: #1694605

mrc0mmand committed:
tree-wide: use 'signed int' instead of 'int' for bit field variables

Suggested by LGTM: https://lgtm.com/rules/1506024027114/

(cherry picked from commit 845d247a3dbca437fd8a118a118328acb322e2c1)

Resolves: #1694605

mrc0mmand committed:
machinectl: check correct FD for a return value

Found by LGTM.

Resolves: #1694605

lnykryn commented:
This pull request **fixes 23 alerts** when merging 368d4f48e517f7765165a56f7a8b11627f334fa8 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-c90f6102cff5c39de985530a061ca8007a2e1c82)

**fixed alerts:**

* 7 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield

---

*Comment posted by [LGTM.com](https://lgtm.com)*

lnykryn commented:
This pull request **introduces 1 alert** and **fixes 25** when merging 300a5ddc3f3f123841929288ad6a7c1e0648f5c9 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-2ffcf7979bde597b4eb569164fbdd71c9f517e63)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 8 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand committed:
shared/install: avoid overwriting 'r' counter with a partial result

We want to store either the first error or the total number of changes in 'r'.
Instead, we were overwriting this with the return value from
install_info_traverse().

LGTM complained later in the loop that:
> Comparison is always true because r >= 0.

Relevant commit: 459500a32cf74edca93ec79ee4cdd9dfcd20eee0

Resolves: #1694605

mrc0mmand committed:
systemctl: set_put() can't return (-)EEXIST

set_put() returns 0 if the key already exists, 1 if the entry
was inserted successfully, and -ENOMEM otherwise.

See: set_put(), hashmap_base_put_boldly()

Found by LGTM.

Resolves: #1694605

lnykryn commented:
This pull request **fixes 26 alerts** when merging 5a5a6768aad8171238913cbc8db0a9f738c06b7a into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-08935a601b81d42e9228fe2bfa17363fc0b8afa3)

**fixed alerts:**

* 9 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

lnykryn commented:
This pull request **fixes 28 alerts** when merging ba2518d9a6ce01a303c89b4c48cf84f1b822c6dd into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-64abe63b5cce3a4c4bac0b92c8cfe7691d4a5b79)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand committed:
lgtm: ignore files dropped in upstream

Resolves: #1694605

mrc0mmand committed:
lgtm: suppress several constant comparisons

Even though LGTM is right is these cases, fixing it would require
substantial refactoring in some cases, so it's better to leave them
here (at least for RHEL 7).

Resolves: #1694605
rhel-only

lnykryn commented:
This pull request **fixes 35 alerts** when merging 70c92469496a831f01ef953442ac8b9c89a050e0 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-3633f76dbe5c6b0fa365769cfff97ac881cb10d8)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Missing variable declaration
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Unneeded defensive code
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand commented:
Even though this PR doesn't fix all identified warnings, I'd review and merge it in its current state, as stuff's getting too hard to follow.

jsynacek commented:
IMHO all the commits need to include "Resolves:" or "Related:", otherwise it won't be possible to merge them downstream.

mrc0mmand commented:
@jsynacek thanks for pointing it out, it completely slipped my mind. Should be fixed now.

lnykryn commented:
This pull request **fixes 35 alerts** when merging 3302ab3e54b53f30bce3ab4a52cc6db62f0edb02 into 25112918d6b54fc1136dde2eef24d6c2d33e55b3 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-914c9183086959ef8bb15ee1b24c7cffa36dfd51)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Missing variable declaration
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Unneeded defensive code
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

jsynacek commented on src/journal/journal-verify.c:
```diff
@@ -70,10 +70,11 @@ static void draw_progress(uint64_t p, usec_t *last_usec) {
-        /* Calculates scale * p / m, but handles m == 0 safely, and saturates */
-
-        if (p >= m || m == 0)
+        if (p >= m || m == 0) // lgtm[cpp/constant-comparison]
```
The original commit used the c-style comment, not the c++-style one. Even though it's really a triviality, I'd say that we should do the same here, as systemd uses the c-style comments everywhere.

mrc0mmand replied:
This is an unfortunate behavior of LGTM - it simply doesn't recognize suppression rules in C++-style comments. I had to do a similar change in upstream as well (see https://github.com/systemd/systemd/commit/1a862e21511918f4f7a27a119624e9629056356b) as the rule was ignored within a C++-style comment.

Excerpt from the [LGTM documentation](https://lgtm.com/help/lgtm/alert-suppression#c-c++):
```
The syntax is language-dependent:

    "// lgtm..." for C/C++, C#, Java and JavaScript
    "* lgtm..." for COBOL
    "# lgtm..." for Python
```

jsynacek commented on src/sysv-generator/sysv-generator.c:
```diff
@@ -568,7 +568,7 @@ static int load_sysv(SysvStub *s) {
                                         }
 
-                                        if (r < 0)
+                                        if (r < 0) // lgtm[cpp/constant-comparison]
```
Were these changes cherry picked from somewhere? If yes, please mention the patch in the commit message. If not, please add "rhel-only" at the end of the commit message.

Also, use the c-style comment, as mentioned in the other patch.

mrc0mmand replied:
Will add "rhel-only" to the commit, as the error-handling part of the `load_sysv()` function was significantly rewritten. Also, for the comment-style, the same explanation applies as above.

jsynacek reviewed (changes requested):
I've found more "problems". Sorry for not finding them earlier and making you rebase again.

lnykryn commented:
This pull request **fixes 7 alerts** when merging 72b47a1d57adf79528dca9dbd9a0b152d181f27e into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-8c6d1bdf91a68e7ba2a1489f4164b6b2682e054a)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

lnykryn commented:
This pull request **fixes 7 alerts** when merging 72b47a1d57adf79528dca9dbd9a0b152d181f27e into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-8c6d1bdf91a68e7ba2a1489f4164b6b2682e054a)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand commented:
Ugh, the LGTM analysis now keeps failing due to compilation error:

```
[2019-04-04 08:45:00] [build] make --no-print-directory all-recursive
[2019-04-04 08:45:00] [build] Making all in .
[2019-04-04 08:45:00] [build]   CC       src/cgls/cgls.o
[2019-04-04 08:45:01] [build] In file included from ./src/shared/util.h:92,
[2019-04-04 08:45:01] [build]                  from ./src/shared/cgroup-show.h:26,
[2019-04-04 08:45:01] [build]                  from src/cgls/cgls.c:29:
[2019-04-04 08:45:01] [build] ./src/shared/missing.h:989:19: error: static declaration of ‘renameat2’ follows non-static declaration
[2019-04-04 08:45:01] [build]  static inline int renameat2(int oldfd, const char *oldname, int newfd, const char *newname, unsigned flags) {
[2019-04-04 08:45:01] [build]                    ^~~~~~~~~
[2019-04-04 08:45:01] [build] In file included from src/cgls/cgls.c:23:
[2019-04-04 08:45:01] [build] /usr/x86_64-linux-gnu/include/stdio.h:164:12: note: previous declaration of ‘renameat2’ was here
[2019-04-04 08:45:01] [build]  extern int renameat2 (int __oldfd, const char *__old, int __newfd,
[2019-04-04 08:45:01] [build]             ^~~~~~~~~
[2019-04-04 08:45:02] [build] make[2]: *** [Makefile:15186: src/cgls/cgls.o] Error 1
[2019-04-04 08:45:02] [build] make[1]: *** [Makefile:19815: all-recursive] Error 1
[2019-04-04 08:45:02] [build] make: *** [Makefile:10424: all] Error 2
```

mrc0mmand commented:
This seems to be a variation of https://github.com/systemd/systemd/issues/8099 fixed by https://github.com/systemd/systemd/commit/5187dd2c403caf92d09f3491e41f1ceb3f10491f, we may need to backport it, somehow...

lnykryn commented:
This pull request **fixes 7 alerts** when merging 38b4fbc5d310e3979804ec3a2f94850eb4e64e91 into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-bb57825c6c6b527ac4ef9d5e3e08fea433268c6b)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand commented:
As the LGTM infrastructure got upgraded, mainly with newer glibc, I had to fix a few other issues:

1) dd5e0d980490fcf394faf3175a13ff555949dc10
This fixes issue mentioned in https://github.com/lnykryn/systemd-rhel/pull/311#issuecomment-479818375. To be somewhat consistent across the `missing.h` file I added the same *workaround* for other syscalls as well, not only the `renameat2`, which was causing issues

2) 3fec580890591c7d4b80efd0ebb2cec7120191dc
Newer glibc stops including `sys/sysmacros.h` from the `sys/types.h`, which causes compilation issues in code using `major()` and `minor()` functions, e.g.:

```
[2019-04-04 09:32:27] [build] /usr/bin/ld: ./.libs/libsystemd-shared.a(libsystemd_shared_la-util.o): in function `get_ctty_devnr':
[2019-04-04 09:32:27] [build] util.c:(.text.get_ctty_devnr+0x8f): undefined reference to `major'
[2019-04-04 09:32:27] [build] /usr/bin/ld: util.c:(.text.get_ctty_devnr+0x12e): undefined reference to `minor'
[2019-04-04 09:32:27] [build] collect2: error: ld returned 1 exit status
```

Let's add it explicitly to files which need it.


With these two additional commits the LGTM analysis should work once again.

lnykryn commented:
This pull request **fixes 7 alerts** when merging 3fec580890591c7d4b80efd0ebb2cec7120191dc into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-6afc2692204bf50d3c557fd8d86dcbb9cbcb24f7)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand commented:
The current LGTM fail is caused by the last two commits not being present in the staging branch, should be fixed once this PR is merged.

![image](https://user-images.githubusercontent.com/679338/55551636-9b9b8d00-56db-11e9-9c84-d623aaddd053.png)

lnykryn commented:
This pull request **fixes 38 alerts** when merging d75397fef3651631869653cab7165ddcd3dfc6b2 into dd004bcc058dbb7c66b8eb1a728732ccfce746c7 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-1233127520b4663870f526aa100c5e4bd6daf890)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Missing variable declaration
* 6 for Assignment where comparison was intended
* 6 for Use of potentially dangerous function
* 4 for Potentially overflowing call to snprintf
* 3 for Implicit downcast from bitfield
* 1 for Unneeded defensive code
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mrc0mmand commented:
Closed in favor of https://github.com/systemd-rhel/rhel-7/pull/2
